v2/internal/reconcilers/arm/arm_client.go

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package arm
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/Azure/azure-service-operator/v2/internal/metrics"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	namespacedSecretName = "aso-credential"
+)
+
+type armClient struct {
+	genericClient  *genericarmclient.GenericClient
+	secretData     map[string][]byte
+	credentialFrom types.NamespacedName
+}
+
+func (c armClient) GenericClient() *genericarmclient.GenericClient {
+	return c.genericClient
+}
+
+func (c armClient) CredentialFrom() types.NamespacedName {
+	return c.credentialFrom
+}
+
+type armClients struct {
+	lock sync.Mutex
+	// clients map hold namespaced name for a secret to all the registered ARM clients
+	clients      map[string]*armClient
+	globalClient *armClient
+}
+
+func NewARMClients(client *genericarmclient.GenericClient, podNamespace string) *armClients {
+	globalClient := &armClient{
+		genericClient:  client,
+		credentialFrom: types.NamespacedName{Name: "aso-controller-settings", Namespace: podNamespace},
+	}
+
+	return &armClients{
+		lock:         sync.Mutex{},
+		clients:      make(map[string]*armClient),
+		globalClient: globalClient,
+	}
+}
+
+func newARMClient(client *genericarmclient.GenericClient, secretData map[string][]byte, credentialFrom types.NamespacedName) *armClient {
+	return &armClient{
+		genericClient:  client,
+		secretData:     secretData,
+		credentialFrom: credentialFrom,
+	}
+
+}
+
+func (c armClients) Register(client *armClient) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.clients[client.CredentialFrom().String()] = client
+}
+
+func (c armClients) Lookup(key string) (*armClient, bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	client, ok := c.clients[key]
+	return client, ok
+}
+
+func (c armClients) GetClient(ctx context.Context, obj genruntime.ARMMetaObject, kubeClient kubeclient.Client, cloud cloud.Configuration) (*armClient, error) {
```

Comment:
I'm wondering if it would make sense to pass `kubeClient` to `armClients` at construction, so that you don't have to pass it every time here?

This could then be: ` GetClient(ctx context.Context, cloud cloud.Configuration, namespace string) (*armClient, error)`

Reply:
Done!